File: bench/benches/deforest.jl
include(joinpath(@__DIR__, "..", "lib", "bench.jl"))

# sum of squares of the odd numbers in [0, N) -- a map/filter/fold pipeline.
# checksum = 4891344686 (< 2^63).
deforest_work() = sum((x * x) % 1000003 for x in 0:19999 if x % 2 == 1)

bench("deforest", deforest_work)

File: bench/benches/deforest.py
import sys, os
sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "lib"))
from bench import bench

# sum of squares of the odd numbers in [0, N) -- a map/filter/fold pipeline.
# map/filter are lazy in py3 (no intermediate lists), but each element still pays
# the per-call lambda overhead -- the interpreter counterpart to ai's fused loop.
# checksum = 1333333330000 (< 2^53).
def work():
    return sum(map(lambda x: x * x, filter(lambda x: x % 2 == 1, range(20000))))

bench("deforest", work)
